### PR TITLE
Supported types in Compute-functions 

### PR DESCRIPTION
**Pull request by jarlebring**
We've discussed this before: what input types should be supported/allowed for the compute functions
```
compute_Mlincomb
compute_Mder
compute_MM
```
Currently, the code is not so consistent, and e.g. `compute_Mlincomb(::DEP,s,V)` only supports `s=eltype(V)` (plus some hacks to support some variants with complex), whereas many other functions work also when s and V are of different type.

For reasons of efficiency, I think it is relevant to have the possibility to have mixed types, e.g., `s` is real and `eltype(V)` is complex. The CPU-time improvement can be seen from this situation: You can multiply a real matrix with complex vector faster than manually doing promotion of the real matrix to a complex matrix:
```
julia> using BenchmarkTools
julia> n=5000; A=randn(n,n); b=randn(n)+1im*randn(n);
julia> Ac=complex(A);
julia> @btime A*b;
  22.771 ms (2 allocations: 78.20 KiB)
julia> @btime Ac*b;
  25.082 ms (2 allocations: 78.20 KiB)
```
It's not much improvement in this case, but I think our interface should not prevent this optimization.

Also: If we say we support mixed types, it should be unit tested. All combinations of all Number-types and compute functions and NEPType.jl-NEP-types.

We do not need to require that a user who defines a new NEP-type always supports all Number-type combinations, but for our core code it would be nice to have carefully handled.

This PR currently contains the unit test for many type-combinations. `compute_Mder` more or less works as expected, but `compute_Mlincomb` fails or returns an incorrect type in several cases.

The PR is not ready yet. Material for discussion.

### CODE REVIEW

**Comment by jarlebring** [comment]
The test now passes for 
```
PEP
DEP
SPMF_NEP
SumNEP
```
for Number-types
```
Float16,Float32,Float64,Complex{Float16},Complex{Float32},Complex{64},BigFloat,Complex{BigFloat}
```
for the compute MM, Mder, Mlincomb. I had to make some exceptions since `exp(:Matrix{BigFloat})` and `exp(:Matrix{Float16})` are not implemented in Julia. 

The combinations gives 1400  tests. They are fast to run, but the JIT-compilation takes considerable time, so I will set a flag `fulltest::Bool` and we do not always need to run all the combinations..

I also set up the possibility to check type stability with `@inferred`. `DEP` was already almost type stable. I disabled type stability check for the other NEP-types.  The other NEP-types need another major changes, which is better to do  in a different PR.

**Comment by jarlebring** [comment]
The main part of this PR is the test. There were not so many changes in the core as I thought, so I think we are pretty safe.

File: src/NEPTypes.jl [hunk 1]
```diff
@@ -306,20 +306,36 @@ Constructor: DEP(AA,tauv) where AA is an array of the
...
+        tauv::Vector{Z}   # the delays (which are always real)
     end
-    function DEP(AA::Array{T,1},tauv::Vector=[0,1.0]) where {T<:AbstractMatrix}
+    function DEP(AA::Vector{T},tauv::Vector=[0,1.0]) where {T<:AbstractMatrix}
         n=size(AA[1],1)
-        tauvconv=Vector{Float64}(tauv);
-        if (real(eltype(AA[1])) != Float64)
-            warning("The delay in a DEP has hardcoded type Float64");
+        if (!isreal(tauv))
+            error("Incorrect construction of DEP. The delays need to be real.")
```
[the remark this answers is not in the document]
Comment (jarlebring) [reply src/NEPTypes.jl]: I also looked for that possibility. I was not able to put that in the struct-definition. The downside now is that one could create a DEP which does not satisfy that if one uses a different constructor than the one we defined.

File: src/NEPTypes.jl [hunk 2]
```diff
@@ -306,20 +306,36 @@ Constructor: DEP(AA,tauv) where AA is an array of the
...
+        end
+
+        tauvconv::Vector{real(eltype(AA[1]))}=Vector{real(eltype(AA[1]))}(tauv);
+
+        if (real(eltype(AA[1]))!=eltype(tauvconv))
```
[the remark this answers is not in the document]
Comment (jarlebring) [reply src/NEPTypes.jl]: Good point. I will remove the if.

File: src/NEPTypes.jl [hunk 3]
```diff
@@ -331,14 +347,14 @@ matrices A_i, and tauv is a vector of the values tau_i
         T=isa(λ,Complex) ? complex(eltype(nep.A[1])) : eltype(nep.A[1]);
 
         if (issparse(nep.A[1])) # Can be determined compiled time since DEP parametric type
-            M=spzeros(T,nep.n,nep.n)
+            M=one(λ)*spzeros(T,nep.n,nep.n)
```
[the remark this answers is not in the document]
Comment (jarlebring) [reply src/NEPTypes.jl]: I cleaned it up a bit, by "computing" the output type.

**Comment by jarlebring** [comment]
@maxbennedich  could you have look at the changes in  `compute_Mder(::SPMF_NEP`?

**Comment by maxbennedich** [comment]
Looks good to me. Eventually we might be able to get rid of the `Zero` field in `SPMF_NEP`, since it seems that it's only referenced in `compute_MM` now.

**Comment by jarlebring** [comment]
I agree. The `compute_MM(::SPMF_NEP` needs quite a bit of rewriting now that we know how to think about types better. I recommend we do it when we make `SPMF_NEP` parametric + type stable.

**Pull request merged**